Lib/typing.py

```diff
@@ -2042,8 +2042,17 @@ class Leaf:
       class Other(Leaf):  # Error reported by type checker
           ...
 
-    There is no runtime checking of these properties.
+    There is no runtime checking of these properties. The decorator
+    sets the ``__final__`` attribute to ``True`` on the decorated object
+    to allow runtime introspection.
     """
+    try:
+        f.__final__ = True
```

Comment:
Thanks for addressing my review Jelle.

I wonder if this will break anyone's code? Should we check for the presence of a `__final__` and bail if we find one? Or am I being overly paranoid here? (I see that `no_type_check` has set its own special attribute without any checks, but I suspect it's been there since the early days).

Reply:
The language reference says that dunder names are reserved for the interpreter, so arguably any breakage is the user's problem, not ours. There are no current references to `__final__` in the source code.

Some searching turned up:
- An old suggestion (https://lists.archive.carbon60.com/python/python/108280 ) to use `__final__` for a runtime-enforced version of `@final`.
- Theano uses an instance attribute `__final__` on one class: https://github.com/Theano/Theano/blob/master/theano/gof/optdb.py#L267.
Neither of those are especially concerning.

But it seems fine to at least skip the assignment if the attribute is already present. Do you think that's valuable?

Reply:
Personally, I think skipping the assignment in some situations would create some weird inconsistencies and needlessly complicate things. I like the super-simple current implementation -- if a user is using dunder attributes, despite the warning in the language reference that there are no backwards-compatibility guarantees for dunder names, I think that's on them :)

Reply:
I'm going to add tests to make sure the current implementation works with common stdlib decorators implemented in C (classmethod, staticmethod, property, lru_cache). Otherwise, I'm voting to stick with the current simple implementation, but I'm happy to change it if there's consensus.

Reply:
That turned up a non-contrived case where this try-except matters: `property` objects don't allow attribute assignment.

Reply:
I'm convinced. Let's leave it as is. Thanks for the research Jelle and Alex.